src/rgw/rgw_rados.cc

```diff
@@ -9158,38 +9162,53 @@ int RGWRados::cls_bucket_list_unordered(RGWBucketInfo& bucket_info,
     " start " << start.name << "[" << start.instance <<
     "] num_entries " << num_entries << dendl;
 
+  static MultipartMetaFilter multipart_meta_filter;
+
   *is_truncated = false;
   librados::IoCtx index_ctx;
 
-  rgw_obj_index_key my_start = start;
-
   map<int, string> oids;
   int r = open_bucket_index(bucket_info, index_ctx, oids, shard_id);
   if (r < 0)
     return r;
   const uint32_t num_shards = oids.size();
 
+  rgw_obj_index_key marker = start;
   uint32_t current_shard;
   if (shard_id >= 0) {
     current_shard = shard_id;
-  } else if (my_start.empty()) {
+  } else if (start.empty()) {
     current_shard = 0u;
   } else {
-    current_shard =
-      rgw_bucket_shard_index(my_start.name, num_shards);
+    std::string oid;
+    if(! multipart_meta_filter.filter(start.name, oid)) {
+      // if multipart_meta_filter fails, must be "regular" and just
+      // use name
+      oid = start.name;
+    }
+    rgw_obj_key obj_key;
+    ceph_assert(rgw_obj_key::parse_raw_oid(oid, &obj_key));
+    if (obj_key.name.empty()) {
+      // if the name is now empty, then the start marker that came in
```

Comment:
could you explain further?  i.e., what made name "now" empty" and why does that mean we should start at the beginning?

Reply:
I think the new explanation is better.